src/porepy/numerics/ad/discretizations.py

```diff
@@ -479,166 +501,292 @@ def _differentiable_mpfa(
-                (1 / inverse_sum_squared, 0), shape=(g.num_faces, g.num_faces)
-            )
-            hf_vals = np.power(t_one_sided.val, -2)
-            half_face_transmissibility_inv_squared = sps.coo_matrix(
-                (hf_vals, (fi, np.arange(sz))), shape=(g.num_faces, sz)
-            ).tocsr()
-            # Face transmissibility differentiated w.r.t. half face transmissibility
-            d_transmissibility_d_t = (
-                face_transmissibility_squared * half_face_transmissibility_inv_squared
-            )
-            # Face half face transmissibility differentiated w.r.t. permeability
-            d_t_d_k = sps.dia_matrix((normals_over_distance, 0), shape=(sz, sz))
+            transmissibility_jac, _ = self._transmissibility(g, global_permeability)
 
+            params = self.gb.node_props(g)[pp.PARAMETERS][self.keyword]
             # Potential for this grid
+            cells_of_grid = self._subdomain_projections.cell_restriction([g]).evaluate(
+                self.dof_manager
+            )
             potential_value = cells_of_grid * potential.val
 
             # Create matrix and multiply into Jacobian
             grad_p = sps.diags(
                 pp.fvutils.scalar_divergence(g).T * potential_value,
                 shape=(g.num_faces, g.num_faces),
             )
-
-            # Compose chain rule for T(t(K(u)). k_one_sided.jac is dK/du.
-            jac = d_transmissibility_d_t * d_t_d_k * k_one_sided.jac
-
             # One half of the product rule applied to (T grad p). The other half
             # is base_flux * potential.jac as added below this loop.
-            grad_p_jac = grad_p * jac
+            grad_p_jac = grad_p * transmissibility_jac
 
             # Boundary conditions
             # Dirichlet values are weighted with the transmissibility on the
             # boundary face.
 
             # Create a copy of the Jacobian at the Dirichlet faces
-            is_dir = self.bc[g]["type"].is_dir
-            jac_dir = sps.csr_matrix(jac.shape)
-            row, col, data = sps.find(jac)
-            active_rows = np.isin(row, np.where(is_dir)[0])
-            # Sign corresponds to gradient
-            jac_dir = sps.csr_matrix(
-                (-data[active_rows], (row[active_rows], col[active_rows])),
-                shape=jac.shape,
-            )
+            is_dir = params["bc"].is_dir
+            is_neu = params["bc"].is_neu
+            # Sign corresponds to gradient under the assumption of outwards
+            # boundary normals
 
             # See tpfa discretization for the following treatment of values
             sort_id = np.argsort(g.cell_faces[is_dir, :].indices)
             bndr_sgn = (g.cell_faces[is_dir, :]).data[sort_id]
-            bc_values = self.bc[g]["values"].copy()
-            bc_values[is_dir] = bndr_sgn * bc_values[is_dir]
+            bc_values = np.zeros(g.num_faces)
+            bc_values[is_dir] = -bndr_sgn * params["bc_values"][is_dir]
```

Comment:
Since I will be confused at some later stage, could you add a short comment why the minus?

Reply:
Does the added comment make any sense?